src/app/modules/form/admin-form/__tests__/admin-form.service.spec.ts

```diff
@@ -1625,4 +1626,167 @@ describe('admin-form.service', () => {
       )
     })
   })
+
+  describe('updateFormLogic', () => {
+    const logicId1 = new ObjectId()
+    const logicId2 = new ObjectId()
+    const mockEmailFormId = new ObjectId()
+    const mockEncryptFormId = new ObjectId()
+
+    const mockFormLogicOld = {
+      form_logics: [
+        {
+          _id: logicId1,
+          logicType: 'showFields',
+        } as ILogicSchema,
+        {
+          _id: logicId2,
+          logicType: 'showFields',
+        } as ILogicSchema,
+      ],
+    }
+
+    const updatedLogic = {
+      _id: logicId1,
+      logicType: 'preventSubmit',
+    } as ILogicSchema
+
+    const mockFormLogicUpdated = {
+      form_logics: [
+        {
+          _id: logicId1,
+          logicType: 'preventSubmit',
+        } as ILogicSchema,
+        {
+          _id: logicId2,
+          logicType: 'showFields',
+        } as ILogicSchema,
+      ],
+    }
+
+    const UPDATE_SPY = jest.spyOn(FormModel, 'updateFormLogic')
+
+    let mockEmailForm: IPopulatedForm,
+      mockEncryptForm: IPopulatedForm,
+      mockEmailFormUpdated: IPopulatedForm,
+      mockEncryptFormUpdated: IPopulatedForm
+
+    beforeEach(() => {
+      mockEmailForm = ({
+        _id: mockEmailFormId,
+        status: Status.Public,
+        responseMode: ResponseMode.Email,
+        ...mockFormLogicOld,
+      } as unknown) as IPopulatedForm
+      mockEncryptForm = ({
+        _id: mockEncryptFormId,
+        status: Status.Public,
+        responseMode: ResponseMode.Encrypt,
+        ...mockFormLogicOld,
+      } as unknown) as IPopulatedForm
+      mockEmailFormUpdated = ({
+        ...mockEmailForm,
+        ...mockFormLogicUpdated,
+      } as unknown) as IPopulatedForm
+      mockEncryptFormUpdated = ({
+        ...mockEncryptForm,
+        ...mockFormLogicUpdated,
+      } as unknown) as IPopulatedForm
+    })
+
+    it('should return ok(updated logic) on successful form logic update for email mode form', async () => {
+      // Arrange
+      const STATIC_UPDATE_SPY = jest
+        .spyOn(FormModel, 'findByIdAndUpdate')
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        .mockReturnValue({
+          exec: jest.fn().mockResolvedValue(mockEmailFormUpdated),
+        })
+
+      // Act
+      const actualResult = await updateFormLogic(
+        mockEmailForm,
+        logicId1.toHexString(),
+        updatedLogic,
+      )
+
+      // Assert
+      expect(actualResult.isOk()).toEqual(true)
+      expect(actualResult._unsafeUnwrap()).toEqual(updatedLogic)
+
+      expect(STATIC_UPDATE_SPY).toHaveBeenCalledWith(
+        mockEmailFormId.toHexString(),
+        {
+          $set: { 'form_logics.$[object]': updatedLogic },
+        },
+        {
+          arrayFilters: [{ 'object._id': logicId1.toHexString() }],
+          new: true,
+          runValidators: true,
+        },
+      )
```

Comment:
I'm not sure this test is behaving as intended. if you've already spied on `FormModel.updateFormLogic`, then `FormModel.findByIdAndUpdate` shouldn't be called, because `updateFormLogic`. could you help me understand how this test works?

Reply:
In this case, I didn't mock the return value of `updateFormLogic`, only checked that it was called with the correct parameters by spying on it. Hence `findByIdAndUpdate` is still called (see below from the static method declaration)

```
// Updates specified form logic.
  FormSchema.statics.updateFormLogic = async function (
    this: IFormModel,
    formId: string,
    logicId: string,
    updatedLogic: ILogicWithId,
  ): Promise<IFormSchema | null> {
    return this.findByIdAndUpdate(
      formId,
      {
        $set: { 'form_logics.$[object]': updatedLogic },
      },
      {
        arrayFilters: [{ 'object._id': logicId }],
        new: true,
        runValidators: true,
      },
    ).exec()
  }
```

Reply:
I see. is it necessary to check the args for `findByIdAndUpdate` as well? seems a bit overkill since we already have tests for `updateFormLogic`. imo it's cleaner to just mock the layer below without digging into the implementation details.

Reply:
edited; no strong views on this